File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:

    bool checkInclusion(string s1, string s2) {
        vector<int> s1_arr(26);
        vector<int> s2_arr(26);
        for (char c: s1){
            s1_arr[c - 'a']++;
        }
        if (s2.length() < s1.length()){
            return false;
        }

        int winStart = 0;
        int winEnd = 0;
        for (int i=0; i<s1.length(); ++i){
            char c = s2[winEnd];
            s2_arr[c - 'a']++;
            winEnd++;
        }
        winEnd--;
        while (winEnd < s2.length()-1){
            if (s1_arr == s2_arr){
                return true;
            } else {
                s2_arr[s2[winStart] - 'a']--;
                winStart++;
                winEnd++;
                s2_arr[s2[winEnd] - 'a']++;
            }
        }
        return s1_arr == s2_arr;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkEqual(int[] origPerm, int[] winString){
        for (int i=0; i<26; i++){
            if (origPerm[i] != winString[i]){
                return false;
            }
        }
        return true;
    }
    public boolean checkInclusion(String s1, String s2) {
        int[] permS1 = new int[26];
        int[] permS2 = new int[26];
        if (s1.length() > s2.length()){
            return false;
        }
        for (int i=0; i<s1.length(); i++){
            permS1[s1.charAt(i) - 'a'] += 1;
            permS2[s2.charAt(i) - 'a'] += 1;
        }
        int winSize = s1.length();
        for (int i=s1.length(); i<s2.length(); i++){
            if (checkEqual(permS1, permS2)){
                return true;
            }
            permS2[s2.charAt(i - winSize) - 'a'] -= 1;
            permS2[s2.charAt(i) - 'a'] += 1;
        }
        return checkEqual(permS1, permS2);


    }
}